src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -2841,6 +2841,50 @@ exprt c_typecheck_baset::do_special_functions(
     overflow.type() = bool_typet{};
     return overflow;
   }
+  else if(identifier == CPROVER_PREFIX "enum_is_in_range")
+  {
+    // Check correct number of arguments
+    if(expr.arguments().size() != 1)
+    {
+      error().source_location = expr.source_location();
+      error() << "__enum_is_in_range expects one argument" << eom;
+      std::ostringstream error_message;
+      error_message << expr.source_location().as_string() << ": "
+                    << "__enum_is_in_range expects one argument, "
+                    << expr.arguments().size() << " were provided";
+      throw invalid_source_file_exceptiont{error_message.str()};
+    }
+    // Replace expr with giant boolean that checks all the values
```

Comment:
I don't think the translation into a condition should be done here, but rather later as a pass on the goto program.

Reply:
+1

Reply:
@peterschrammel Could you elaborate where exactly you'd like this to happen?

Reply:
I think it would fit well into the goto_check pass. Then you can even reuse the logic for generating the conditons that is already there.

Reply:
@peterschrammel This should work without the pass being enabled as it's more generally useful though.

Reply:
@peterschrammel It would be nice to be able to reuse the same check logic for both kinds of check. However we don't think it fits into `goto_check` because it is replacing an existing expression rather than adding new checks. Would you be OK with us adding it in `src/goto-programs/builtin_functions.cpp`, function `goto_convertt::do_function_call_symbol`? This seems to be the place where the other built in functions are implemented / replaced.

Reply:
`builtin_functions` is certainly a better place than `c_typecheck_expr`.

Reply:
@peterschrammel This is addressed now.